client/mm/event_log.go

```diff
@@ -0,0 +1,672 @@
+	return db.Update(func(tx *bbolt.Tx) error {
+		botRuns := tx.Bucket(botRunsBucket)
+		key := runKey(startTime, mkt)
+		runBucket := botRuns.Bucket(key)
+		if runBucket == nil {
+			return fmt.Errorf("nil run bucket for key %x", key)
+		}
+
+		return runBucket.Put(endTimeKey, encode.Uint64Bytes(uint64(endTime)))
+	})
+}
+
+// storeEvent stores/updates a market making event.
+func (db *boltEventLogDB) storeEvent(startTime int64, mkt *MarketWithHost, e *MarketMakingEvent) {
+	db.updateEvent <- &eventUpdate{
+		runKey: runKey(startTime, mkt),
+		e:      e,
+	}
+}
+
+// runEvents returns the events that took place during a run. If n == 0, all of the
+// events will be returned. If refID is not nil, the events including and after the
+// event with the ID will be returned. If pendingOnly is true, only pending events
+// will be returned.
+func (db *boltEventLogDB) runEvents(startTime int64, mkt *MarketWithHost, n uint64, refID *uint64, pendingOnly bool) ([]*MarketMakingEvent, error) {
+	events := make([]*MarketMakingEvent, 0, 32)
+	return events, db.View(func(tx *bbolt.Tx) error {
+		botRuns := tx.Bucket(botRunsBucket)
+		key := runKey(startTime, mkt)
+		runBucket := botRuns.Bucket(key)
+		if runBucket == nil {
+			return fmt.Errorf("nil run bucket for key %x", key)
+		}
+
+		eventsBkt := runBucket.Bucket(eventsBucket)
+		if eventsBkt == nil {
+			return nil
+		}
+
+		cursor := eventsBkt.Cursor()
+		var k, v []byte
+		if refID == nil {
+			k, v = cursor.Last()
+		} else {
+			k, v = cursor.Seek(encode.Uint64Bytes(*refID))
+		}
+
+		for ; k != nil; k, v = cursor.Prev() {
+			ver, pushes, err := encode.DecodeBlob(v)
+			if err != nil {
+				return err
+			}
+			if ver != 0 {
+				return fmt.Errorf("unknown version %d", ver)
+			}
+			if len(pushes) != 1 {
+				return fmt.Errorf("expected 1 push for event, got %d", len(pushes))
+			}
+
+			var e MarketMakingEvent
+			err = json.Unmarshal(pushes[0], &e)
+			if err != nil {
+				return err
+			}
+
+			if pendingOnly && !e.Pending {
+				continue
+			}
+
+			events = append(events, &e)
+			if n > 0 && uint64(len(events)) >= n {
+				break
+			}
+		}
+
+		return nil
+	})
+}
+
+// updateFiatRates updates the fiat rates for a run. This should be updated
+// periodically during a run to have accurate P/L information.
+func (db *boltEventLogDB) updateFiatRates(startTime int64, mkt *MarketWithHost, fiatRates map[uint32]float64) error {
+	return db.Update(func(tx *bbolt.Tx) error {
+		botRuns := tx.Bucket(botRunsBucket)
+		key := runKey(startTime, mkt)
+		runBucket := botRuns.Bucket(key)
+		if runBucket == nil {
+			return fmt.Errorf("nil run bucket for key %x", key)
+		}
+
+		fiatRatesBkt, err := runBucket.CreateBucketIfNotExists(fiatRatesBucket)
+		if err != nil {
+			return err
+		}
+		for assetID, rate := range fiatRates {
+			fiatRatesBkt.Put(encode.Uint32Bytes(assetID), encode.Uint64Bytes(fiatToUint(rate)))
+		}
+
+		return nil
+	})
+}
```

Comment:
Shouldn't we be storing fiat rates associated with both the start and end of the run?

Reply:
I was storing the rates at the end for the profit calculation. Why store the rates at the start?